cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/search/AzureSearch.scala

```diff
@@ -304,25 +309,45 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with VectorCol
         UDFUtils.oldUdf(checkForErrors(fatalErrors) _, ErrorUtils.ErrorSchema)(col("error"), col("input")))
   }
 
+  private def assertNoNestedVectors(fields: Seq[IndexField]): Unit = {
+    def checkVectorField(field: IndexField): Unit = {
+      if (field.dimensions.nonEmpty && field.vectorSearchConfiguration.nonEmpty) {
+        throw new IllegalArgumentException(s"Nested field ${field.name} is a vector field, vector fields in nested" +
+          s" fields are not supported.")
+      }
+      field.fields.foreach(_.foreach(checkVectorField))
+    }
+    fields.foreach(checkVectorField)
+  }
+
   private def getVectorColNameTypeTupleFromIndexJson(indexJson: String): Seq[(String, String)] = {
     parseIndexJson(indexJson).fields
       .filter(f => f.vectorSearchConfiguration.nonEmpty && f.dimensions.nonEmpty)
       .map(f => (f.name, f.`type`))
   }
   private def castDFColsToVectorCompatibleType(vectorColNameTypeTuple: Seq[(String, String)],
                                                df: DataFrame): DataFrame = {
-
     vectorColNameTypeTuple.foldLeft(df) { case (accDF, (colName, colType)) =>
-      assert(accDF.columns.contains(colName), s"Column $colName not found in dataframe columns ${accDF.columns.toList}")
-      val colDataType = accDF.schema(colName).dataType
-      assert(colDataType match {
-        case ArrayType(elementType, _) => elementType == FloatType || elementType == DoubleType
-        case VectorType => true
-        case _ => false
-      }, s"Vector column $colName needs to be one of (ArrayType(FloatType), ArrayType(DoubleType), VectorType)")
-      accDF.withColumn(colName, accDF(colName).cast(edmTypeToSparkType(colType, None)))
+      if (!accDF.columns.contains(colName)) {
+        println(s"Column $colName is specified in either indexJsonOpt or vectorColsOpt but not found in dataframe " +
+          s"columns ${accDF.columns.toList}")
+        accDF
+      }
+      else {
+        val colDataType = accDF.schema(colName).dataType
+        assert(colDataType match {
+          case ArrayType(elementType, _) => elementType == FloatType || elementType == DoubleType
+          case VectorType => true
+          case _ => false
+        }, s"Vector column $colName needs to be one of (ArrayType(FloatType), ArrayType(DoubleType), VectorType)")
+        if (colDataType.isInstanceOf[ArrayType]) {
+          accDF.withColumn(colName, accDF(colName).cast(edmTypeToSparkType(colType, None)))
+        } else {
+          accDF.withColumn(colName, vector_to_array(accDF(colName), edmTypeToVectordType(colType)))
```

Comment:
For Vector types, I learned that we can't use cast, we need to use vector_to_array function, which only expects either "float32" or "float64" as dtype: https://github.com/apache/spark/blob/v3.4.1/mllib/src/main/scala/org/apache/spark/ml/functions.scala#L63
I couldn't find an existing function that would map FloatType to float32 for example, so I wrote my own clunky function. Maybe the better thing to do is to omit the dtype, which defaults to float64, and let ACS do the casting when we send it.